### DISCUSSION TITLE
Running Breeze Consumes 100% Host Memory

### DISCUSSION THREAD

**Discussion by calebstewart** [apache/airflow#29731] [Q&A] [answered]
I just cloned the repository today (latest `main`) and was following the [Contributor Quick Start](https://github.com/apache/airflow/blob/main/CONTRIBUTORS_QUICK_START.rst), but if I try to run `breeze start-airflow`, then each of the `triggerer`, `scheduler` and `webserver` processes attempt to consume 100% of my host machine's memory.

For some context, I'm not running in a constrained environment. On the contrary, I have 48 cores, 96GB of RAM and well over a TB of disk space available on my host machine. As soon as any of the main 3 processes mentioned above starts, the memory usage starts climbing rapidly, and then the kernel OOM steps in and kills them. No services ever start, and normally all three get killed at around ~60GB of allocated memory by the kernel OOM reaper. Occasionally, two of the processes will get killed and the third (e.g. the webserver) might keep running. In those cases, I've seen the web server continue running and consuming ~70-80GB of RAM, but it's not really usable since the other components can't start.

I did try to search through issues (open/closed) and discussions but didn't find anything identical to my situation. I found a few memory leak discussions but those were over longer time periods and related to K8s or other deployment scenarios. I'm explicitly just trying to get the development environment running.

```sh
$ docker --version
Docker version 23.0.1, build a5ee5b1dfc

$ docker-compose --version
Docker Compose version 2.16.0
```

![image](https://user-images.githubusercontent.com/7529189/221026432-594341c3-79e1-4d0d-a88d-61249d2c35f1.png)

**Comment by potiuk** [comment]
This is rather strange and not easily reproducible - and it might likely indicate your docker installation is broken somehow. 
When I start Breeze on my machine wiht `breeze start-airflow` I got compleletely steady ~1.5 GB Use (this is with Postgres as backend):

Result of `docker stats`:

<img width="1168" alt="Screenshot 2023-02-24 at 10 07 46" src="https://user-images.githubusercontent.com/595491/221138060-e3f92579-879b-42b9-8b77-650a4d6e11dc.png">

I run it for 20 minutes without even the least significant digit change once it hit 1.523 GB.

What I would suggest is to go step-by-step through the points in https://github.com/apache/airflow/blob/main/BREEZE.rst#troubleshooting and see whether they fix your problems.

Also it might be worthwile to check if running some other docker run software (with mounted volumes) has the same effects. 

Few other issues that might make your machine having troubles is some low-level kernel or filesystem related problems. It could be for example thet the filesystem you try to check-out Airflow on is somewhat unfriendly with docker volume mounts or that your docker engine uses some non-standard settings for backing storage (docker engine might use various backend storages to store theyr layers and containers. It might for example be that the drivers used by docker have memory leak and they are causing memory growth by synchronising volumes. 

Or it might be a kernel issue in your machine - so maybe rather than looking for similar "Airflow" issues - look for similar "docker" issues for your machine.

Also If you run your docker in a VM on Windows / WSL2 or others, you might have various problems related to particular issues related with those (and filesystem issues there are usually one of the culprits of strange behaviours). Making sure that your filesystem where you checkout the sources is "native" (ie. not additionally mounted between HOST OS and VM that runs docker engine) usually helps  a lot.

**Comment by j2cry** [comment]
I have the same problem when trying to run in docker on Fedora. After some experimentation, I found that using any version starting from 2.3.4 results in what you described. While on Ubuntu 20.04 everything works fine.

**Reply by potiuk** [reply]
Very strange. Can anyone who experiences it posts some logs of what's happening with airflow processes when it happens? Ideally you could try to do what airflow does when running start-airflow:

1) run  `breeze` (just the shell)
2) run `tmux`
3) split terminal into 4 (ctl + b followed by " or % few times)
4) run "airflow scheduler", "airflow triggerer", "airflow webserver" one-by-one 

and see if any of those causes the memory growth ? And maybe dump processes that actually start consuming memory ?

**Reply by j2cry** [reply]
I'm running airflow using the official [docker-compose.yaml](https://airflow.apache.org/docs/apache-airflow/2.5.1/docker-compose.yaml).
Also, tried to comment out the services: scheduler, worker, triggerer, webserver. Both individually and in different combinations, but this takes no effect. If any of those starts it consumes all memory and cpu.
You can try to reproduce this in VM.

Here's the tail of `docker compose up`:
```
airflow-airflow-init-1       | User "airflow" created with role "Admin"
airflow-airflow-init-1       | 2.5.1
airflow-airflow-init-1 exited with code 0
airflow-airflow-worker-1     | 
airflow-airflow-scheduler-1  | 
airflow-airflow-triggerer-1  | 
airflow-airflow-webserver-1  | 
airflow-airflow-scheduler-1  | BACKEND=redis
airflow-airflow-scheduler-1  | DB_HOST=redis
airflow-airflow-worker-1     | BACKEND=redis
airflow-airflow-worker-1     | DB_HOST=redis
airflow-airflow-worker-1     | DB_PORT=6379
airflow-airflow-scheduler-1  | DB_PORT=6379
airflow-airflow-worker-1     | 
airflow-airflow-scheduler-1  | 
airflow-airflow-webserver-1 exited with code 137
airflow-airflow-triggerer-1 exited with code 137
airflow-airflow-worker-1 exited with code 137
airflow-airflow-scheduler-1 exited with code 137
```

**Reply by potiuk** [reply]
Does it work the same when you start airlfow on any of those without docker? What docker configuration do you use? How did you install docker?

**Reply by j2cry** [reply]
Just checked in GNOME Boxes: on Ubuntu 20.04/22.04 everything works even with 2 cores 4 Gb RAM, on Fedora it doesn't.
I use default docker installed according to the [manual](https://docs.docker.com/engine/install/fedora/) (Install using the repository) and I have not tried to run without it.

**Reply by potiuk** [reply]
OK. Thanks!. Very interesting.

**Reply by potiuk** [reply]
Trying to reproduce it now.

**Reply by j2cry** [reply]
Tried to run without docker with the following configuration:
Gnome BOXES, 2 cores 5Gb RAM, python3.9 virtual environment: airflow 2.5.1 + psycopg2-binary + redis. PostgreSQL and Redis are still in docker and initialized with `docker compose up airflow-init`. Starting services one by one and ... everything's OK.

![tmux](https://user-images.githubusercontent.com/48529474/221238054-3c57b8bd-33af-43bb-abe3-0777c39029a0.png)
P.S. thanx for `tmux` - i've never used it before.

**Reply by potiuk** [reply]
So is it just 'breeze start-airlfow" / docker-compose up that causes the problem?

**Reply by j2cry** [reply]
In my case, obviously, the problem is somehow related to docker and Fedora. I'll try to build the image manually and see if it will work.

**Reply by potiuk** [reply]
Yep I reproduced in fedora's VM

**Reply by potiuk** [reply]
Ok. This is a problem which is explained here (airflow and it *not* working as of 2.3.4 is actually mentioned there)   https://github.com/moby/moby/issues/43361

The underlying problem is that ontainerd for Fedora kernel set LimitNOFILE to infinity for containerd and likely some of the 650 dependencies of airflow  that is installed together with airflow has a problem with it: https://github.com/containerd/containerd/pull/7566#issuecomment-12854173. There are other dockerized apps failing in the same way (mysql seems to have the same problem). 

Luckily the https://github.com/moby/moby/issues/43361#issuecomment-1227617516  has a solution. Instead of using fedora-maintained containerd/docker, install Docker Desktop.  That solves the problem as the containerd settings are sane there apparently.

**Reply by potiuk** [reply]
PR with documentation update describing the workaround in: https://github.com/apache/airflow/pull/29751

**Reply by j2cry** [reply]
Wow. Thanks!

**Reply by calebstewart** [reply]
Awesome, that's an easy fix. I appreciate y'all going through this in my absence. Glad we found a quick resolution!

**Reply by rodrigorochag** [reply]
May I ask how you did for get it working? I am having the same trouble here. I'm using Airflow 2.5.1 on Docker-compose Fedora 37.

I found a solution: https://github.com/moby/moby/issues/43361

**Reply by Blizzke** [reply]
> Luckily the [moby/moby#43361 (comment)](https://github.com/moby/moby/issues/43361#issuecomment-1227617516) has a solution. Instead of using fedora-maintained containerd/docker, install Docker Desktop. That solves the problem as the containerd settings are sane there apparently.

Or update your entrypoint (add the `ulimit` just before the last line):  
```
ulimit -n 1024
exec "airflow" "${@}"
```
This also keeps things in check (coming from a 0-to-64+GB-in-10-seconds situation here)

There might be other places to run this without updating your airflow image, but since I was already doing that this was the least effort.

**Reply by calebstewart** [reply]
The easiest fix for me was to just make the change from containerd/containerd#7566 locally using an systemd override file. That can be easily undone later if there is an upstream fix, and doesn't actually require any changes to any docker images. You can use the systemctl command to add an override like:

```sh
sudo systemctl edit containerd.service
```

Or create the file `/etc/systemd/system/containerd.service.d/override.conf`. Either way, place this content in it:

```ini
[Service]
LimitNOFILE=1048576
```

This makes everything work perfectly. :+1: 

edit: if you choose to edit the `override.conf` directly, I believe you'll need to run `systemctl daemon-reload` to reload the unit files. I used `systemctl edit` and didn't need to, but I think that command does some magic internally to auto-reload that unit for you.

**Reply by calebstewart** [reply]
You'll also need to restart the service if it's currently running obviously.

**Reply by Blizzke** [reply]
> This makes everything work perfectly. +1

Didn't even think about that, darn. 
That is a much more elegant solution, thank you for that!

**Reply by potiuk** [reply]
Cool to see all those workarounds, well ... working :). It's a pity containerd has not yet realised how much of a problem it is.

**Comment by calebstewart** [comment, marked as answer]
I don't see a way to mark one of the comments as the "answer" here, so I'll re-post this for clarity.

The underlying problem is that `containerd` made a change which sets the `nofile` limit to `infinity` on the service. This trickles down to all the child containers. Some services such as MySQL and apparently something within Airflow attempt to pre-allocate resources based on the number of potentially open files. Therefore, when the limit is set to infinity, they allocate huge amounts of memory on startup in an attempt to pre-optimize.

All of this together means that if you have a relatively new version of `containerd`, then certain container images will cause memory exhaustion as described in this discussion. There is an open PR over at containerd/containerd#7566 which reverts the change and sets the `LimitNOFILE` property on the `containerd.service` systemd unit back to `1048576` which was the old default value. This fix seems to stop the runaway memory problem, but has not been merged.

In https://github.com/moby/moby/issues/43361#issuecomment-1227617516, a  user claimed that using Docker Desktop for Linux does not have this problem. This is likely due to it using an older version of `containerd`. This can work for you, if you want to use that distribution of docker, and don't mind reinstalling.

Alternatively, @Blizzke mentioned that he  could get around this by calling `ulimit` at runtime in the container before starting the offending services. This should work, but requires modifying your container images, and may not be suitable for everyone.

**My *personal* preferred solution**: A suitable workaround is just to manually override the limit set in the systemd unit. Systemd allows you to provide an override file which, as the name implies, overrides select properties in a unit without having to modify the source unit (see Example 2 in the [systemd unit man page](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Examples)). This has the benefit of not requiring modifications to any container images, and not uninstalling or reinstalling any software, and should not be affected by updating `containerd` in the future. Reverting this change is as simple as removing the override file.

To create an override file, use the command `sudo systemctl edit containerd.service`. This should launch an editor (normally `vim`) where you can add your override configuration. To fix this issue, I used the following content:

```ini
[Service]
LimitNOFILE=1048576
```

This command will create a file named `/etc/systemd/system/containerd.service.d/override.conf`. You should then restart your service. If you manually create the file or don't see the changes, you can use the `systemctl daemon-reload` command to explicitly have systemd reload unit files from disk, which will trigger it to recognize your override file.

To remove this modification in the future (when the problem is hopefully fixed upstream), you can simply remove the override file.

**Comment by potiuk** [comment]
Looks like we know the dependency that causes it now - it's python-daemon so maybe we will be able to do something about it (looking at it now). See https://github.com/apache/airflow/issues/29841#issuecomment-1450419914 and the issue opened in https://pagure.io/python-daemon/issue/72 (upvoting might increase our chances to get it solved upstream in python-daemon)

**Comment by potiuk** [comment]
@calebstewart @Blizzke @j2cry  @rodrigorochag  I'd really love if you could apply the two commits / patch airflow to see if that fixes the problem regardless from the workaround https://github.com/apache/airflow/pull/29848

**Reply by calebstewart** [reply]
I'll give that branch a go tomorrow without the workaround :+1:

**Reply by potiuk** [reply]
Seems that it might be not needed  - see the latest comment in https://pagure.io/python-daemon/issue/72 - the new "main" version of `python-daeamon` seems to fix the issue (I will test myself and follow up and ask to speed up the release.

**Comment by potiuk** [comment]
The python-daemon==3.0.0 has been releassed with the fix - you can upgrade it and it should fix the problem. I am adding >= 3.0.0 for 2.5.2 release as well.
